compiler/runtime/OMRRelocationTarget.hpp

```diff
@@ -0,0 +1,100 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef OMR_RELOCATION_TARGET_INCL
+#define OMR_RELOCATION_TARGET_INCL
+
+#ifndef OMR_RELOCATION_TARGET_CONNECTOR
+#define OMR_RELOCATION_TARGET_CONNECTOR
+namespace OMR { class RelocationTarget; }
+namespace OMR { typedef OMR::RelocationTarget RelocationTargetConnector; }
+#endif
+
+
+#include <stddef.h>
+#include <stdint.h>
+#include "env/TRMemory.hpp"
+#include "env/jittypes.h"
+#include "env/JitConfig.hpp"
+
+namespace TR {
+   class OpaqueClassBlock;
+   class RelocationRecord;
+   class RelocationRuntime;
+   class RelocationTarget;
+}
+/**
+ * TR_RelocationTarget defines how a platform target implements the individual 
+ * steps of processing relocation records.
+ * This is intended to be a base class that should not be itself instantiated
+ */
+namespace OMR{
+class  OMR_EXTENSIBLE RelocationTarget
```

Comment:
If this class is extensible, I don't don't know if having all these virtual methods is the way to go. On OpenJ9, each platform extends `TR_RelocationTarget` to override some of these methods. However, if if the proposed `TR::RelocationTarget` class here is extensible, then extending projects should extend it using extensible classes. Otherwise, if we're going to have virtual methods, we should not make this class extensible. @0xdaryl / @mstoodle  do you have an opinion on this?

Reply:
depends who is doing the overriding of the virtual methods....if `J9::RelocationTarget` is the one providing implementations of the "virtual" functions to override the implementations in `OMR::RelocationTarget` then that's probably not right. But if it's subclasses of `TR::RelocationTarget` then probably not.

Here, it depends on whether the platform relocation target classes are all part of `TR::RelocationTarget` or are they subclasses of `TR::RelocationTarget`. All the uses of `TR::RelocationTarget` in other files suggest to me the former, which means they shouldn't be virtual. But I don't actually see the platform implementations of `RelocationTarget` in this PR....did I miss them somewhere?

Reply:
Hi @mstoodle, thank you for your comment. I have not included them as part of this PR, since I am not familiar with platform specific implementations of relocation targets, so I have proposed the ones I know of to be generic ( [X86RelocationTarget](https://github.com/eclipse/openj9/blob/master/runtime/compiler/x/runtime/X86RelocationTarget.cpp)  being an inspiration), with possibility of overriding by an extensible class. Is that okay?

Reply:
maybe. you should at least have a notion for how you expect different relocation targets to be implemented though (as virtual subclasses or as part of the extensible class composition). That's the fundamental question Irwin's asked. The relocation targets in OpenJ9 were created before we had extensible classes, so they necessarily used virtual overrides even though only one relocation target was used in any one run. I therefore suspect different targets would be implemented as architecture specializations for the extensible class.

which is a long way of saying: these methods probably should not be virtual.

Reply:
Fixed, with all other files as well!